internal/civisibility/integrations/gotesting/coverage/test_coverage.go

```diff
@@ -0,0 +1,436 @@
+	if err != nil {
+		log.Debug("civisibility.coverage: error creating temporary directory: %v", err)
+	} else {
+		log.Debug("civisibility.coverage: temporary coverage directory created: %s", temporaryDir)
+	}
+	integrations.PushCiVisibilityCloseAction(func() {
+		_ = os.RemoveAll(temporaryDir)
+	})
+
+	// executing go list -f '{{.Module.Path}};{{.Module.Dir}}' to get the module path and module dir
+	stdOut, err := exec.Command("go", "list", "-f", "{{.Module.Path}};{{.Module.Dir}}").CombinedOutput()
+	if err != nil {
+		log.Debug("civisibility.coverage: error getting module path and module dir: %v", err)
+	} else {
+		parts := strings.Split(string(stdOut), ";")
+		if len(parts) == 2 {
+			modulePath = strings.TrimSpace(parts[0])
+			moduleDir = strings.TrimSpace(parts[1])
+		}
+	}
+}
+
+// CanCollect returns whether coverage can be collected.
+func CanCollect() bool {
+	return mode == "count" || mode == "atomic"
+}
+
+// GetCoverage returns the total coverage percentage for the test package
+func GetCoverage() float64 {
+	if !CanCollect() {
+		return 0
+	}
+
+	restore := setStdOutToTemp()
+	coverageFile := filepath.Join(temporaryDir, "global_coverage.out")
+	_, err := tearDown(coverageFile, "")
+	restore()
+	if err != nil {
+		log.Debug("civisibility.coverage: error getting coverage file: %v", err)
+	}
+
+	defer func(cFile string) {
+		err = os.Remove(cFile)
+		if err != nil {
+			log.Debug("civisibility.coverage: error removing coverage file: %v", err)
+		}
+	}(coverageFile)
+
+	totalStatements, coveredStatements, err := getCoverageStatementsInfo(coverageFile)
+	if err != nil {
+		log.Debug("civisibility.coverage: error parsing coverage file: %v", err)
+	}
+
+	if totalStatements == 0 {
+		return 0
+	}
+
+	return float64(coveredStatements) / float64(totalStatements)
+}
+
+// NewTestCoverage creates a new test coverage.
+func NewTestCoverage(sessionID, moduleID, suiteID, testID uint64, testFile string) TestCoverage {
+	testFile = utils.GetRelativePathFromCITagsSourceRoot(testFile)
+	return &testCoverage{
+		sessionID: sessionID,
+		moduleID:  moduleID,
+		suiteID:   suiteID,
+		testID:    testID,
+		testFile:  testFile,
+	}
+}
+
+// CollectCoverageBeforeTestExecution collects coverage before test execution.
+func (t *testCoverage) CollectCoverageBeforeTestExecution() {
+	if !CanCollect() {
+		return
+	}
+
+	restore := setStdOutToTemp()
+	t.preCoverageFilename = filepath.Join(temporaryDir, fmt.Sprintf("%d-%d-%d-pre.out", t.moduleID, t.suiteID, t.testID))
+	_, err := tearDown(t.preCoverageFilename, "")
+	restore()
+	if err != nil {
+		log.Debug("civisibility.coverage: error getting coverage file: %v", err)
+	}
+}
+
+// CollectCoverageAfterTestExecution collects coverage after test execution.
+func (t *testCoverage) CollectCoverageAfterTestExecution() {
+	if !CanCollect() {
+		return
+	}
+
+	restore := setStdOutToTemp()
+	t.postCoverageFilename = filepath.Join(temporaryDir, fmt.Sprintf("%d-%d-%d-post.out", t.moduleID, t.suiteID, t.testID))
```

Comment:
why is this line inside the `restore := setStdOutToTemp()` block? isn't it only necessary for the `tearDown` fn to be in that block?

Reply:
True, I did a refactor to remove that method and just create a tearDown wrapper in [1ba2dda](https://github.com/DataDog/dd-trace-go/pull/2943/commits/1ba2ddaeb7768e4f9cf9208d81a6db7890cdaed0)